include/core/rtos/BytePoolBase.hpp

```diff
@@ -0,0 +1,84 @@
+
+#ifndef _EVT_RTOS_BYTEPOOL_
+#define _EVT_RTOS_BYTEPOOL_
+
+#include <core/rtos/Enums.hpp>
+#include <cstdint>
+#include <tx_api.h>
+
+namespace core::rtos {
+
+/**
+ * A pure virtual class that allows us to pass the BytePool template class into
+ * external methods without the compiler complaining.
+ */
+class BytePoolBase {
+public:
+    /**
+     * Allocates memory from the BytePool and returns a pointer to the start of it.
+     *
+     * @param[in] amount how much memory (in bytes) that is requested.
+     * @param[out] memoryPointer a reference to the pointer to the block of memory that is requested.
+     * @param[in] waitOption How long (in ticks) the calling thread should wait for the memory to become available:
+     * 0 for no wait, and TX_WAIT_FOREVER for waiting forever.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    virtual TXError allocateMemory(std::size_t amount, void** memoryPointer, uint32_t waitOption) = 0;
+
+    /**
+     * Initializes the bytepool within the threadx kernel.
+     */
+    virtual TXError init() = 0;
+
+    /**
+     * Releases the memory at the given pointer from the BytePool.
+     *
+     * @param[in] memoryPointer pointer to the memory to release.
+     * @return The first error found by the function (or Success if there was no error).
+     */
+    virtual TXError releaseMemory(void* memoryPointer) = 0;
+
+    /**
+     * Retrieves the number of unallocated bytes left in the BytePool.
+     *
+     * @param[out] availableBytes a pointer to a place to store the number of available bytes.
```

Comment:
Yeah... I think it's still interpretable as is. I understand your point and I probably would not do it this way if I was to remake this completely, but I think the amount of work involved in refactoring this is not worth the outcome, especially because it is still clear in it's meaning.

Reply:
I think it is worth making the distinction when you're just using a pointer so you can get data out and when you're using a pointer to get data out. As a reader of your documentation, it's really annoying to parse through "a pointer to a place to store..." over and over again. I know it's going to be annoying to fix, but the RTOS is a really important, central thing, and I would like it to be as clean as possible.

If @mjh9585 disagrees, I would be willing to concede this though

Reply:
While I agree that it should be unnecessary to specify that it is a pointer that you are returning data in, it can be confusing to someone that doesn't have a solid understanding of pointers or pointer design patterns. I would say maybe shortening it to "The returned number of available bytes." since people understand the concept of returning as giving a value back.

Reply:
As a side note, a lot of these pointers should really be references instead to insure that null cannot be passed in and improve safety.

Reply:
I would argue that the `out` expresses that the value is returned, but I would agree to that compromise

Reply:
I agree with the naming thingy. Weirdly the reference isn't actually necessary, the way the underlying threadx method works is that if any of the passed in pointers are null it just assumes you didn't want that information, so if they do pass in null it will just do nothing. I can make them references in case, but I think for stuff like the char** it will be weird for it to be a char*&.

Reply:
Reworded them to be "The returned [thing]" instead of the weird pointer stuff. Did not change to references for the above reasons.